cranelift/isle/isle/src/serialize.rs

```diff
@@ -0,0 +1,823 @@
+                HasControlFlow::Equal(source, _) => source,
+                HasControlFlow::Loop(source) => source,
+            };
+            score.state = self.scope.ready[source.index()];
+
+            // Candidates which have already been matched in this partition
+            // must not be matched again. (Note that equality constraints are
+            // de-duplicated using more complicated equivalence-class checks
+            // instead.)
+            if score.state == BindingState::Matched {
+                return false;
+            }
+
+            // The sort key is not based solely on how many rules have this
+            // constraint, but on how many such rules can go into the same
+            // block without violating rule priority. This number can grow
+            // as higher-priority rules are removed from the partition, so
+            // we can't drop candidates just because this is zero. If some
+            // rule has this constraint, it will become viable in some later
+            // partition.
+            let partition = kind.partition(self.rules, order);
+            score.count = partition.valid;
+
+            // Only consider constraints that are present in some rule in the
+            // current partition. Note that as we partition the rule set into
+            // smaller groups, the number of rules which have a particular
+            // kind of constraint can never grow, so a candidate removed here
+            // doesn't need to be examined again in this partition.
+            partition.any_matched
+        };
+
+        // Remove false candidates, and recompute the candidate sort key for
+        // the current set of rules in `order`.
+        self.scope
+            .candidates
+            .retain_mut(|candidate| set_score(&mut candidate.score, candidate.kind.0));
+
+        // Find the best normal candidate.
+        let mut best = self.scope.candidates.iter().max().cloned();
+
+        // Equality constraints are more complicated. We need to identify
+        // some pair of binding sites which are constrained to be equal in at
+        // least one rule in the current partition. We do this in two steps.
+        // First, find each single binding site which participates in an
+        // equality constraint in some rule. We compute the best-case `Score`
+        // we could get, if there were another binding site where all the rules
+        // constraining this binding site require it to be equal to that one.
+        self.scope.equal_candidates.retain_mut(|candidate| {
+            let source = candidate.source.0;
+            // The `Equal` case checks that both binding sites are in the same
+            // equivalence class. By specifying the same binding site twice,
+            // we're checking whether the binding site is in any equivalence
+            // class at all.
+            set_score(&mut candidate.score, HasControlFlow::Equal(source, source))
```

Comment:
What do you think about introducing a function on `HasControlFlow` called something like `any_equiv_class` that constructs the `Equal(source, source)` case? Something like:
```rust
impl HasControlFlow {
  fn any_equivalence_class(source: BindingId) -> Self {
    Self::Equal(source, source)
  }
}
```
It might help when seeing uses where the args are the same.